src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,294 @@
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * db.lock.update(
+     *     {
+     *         _id: "test",
+     *         expires_at: {
+     *             $lte : new Date()
+     *         }
+     *     },
+     *     {
+     *         _id: "test",
+     *         token: {# unique token #},
+     *         acquired: new Date(),
+     *         expires_at: new Date({# now + ttl #})
+     *     },
+     *     {
+     *         upsert: 1
+     *     }
+     * );
+     */
+    public function save(Key $key)
+    {
+        $expiry = $this->createDateTime(microtime(true) + $this->initialTtl);
+
+        $filter = array(
+            $this->options['resource_field'] => (string) $key,
+            '$or' => array(
+                array(
+                    $this->options['token_field'] => $this->getToken($key),
+                ),
+                array(
+                    $this->options['expiry_field'] => array(
+                        '$lte' => $this->createDateTime(),
+                    ),
+                ),
+            ),
+        );
+
+        $update = array(
+            '$set' => array(
+                $this->options['resource_field'] => (string) $key,
+                $this->options['token_field'] => $this->getToken($key),
+                $this->options['acquired_field'] => $this->createDateTime(),
+                $this->options['expiry_field'] => $expiry,
+            ),
+        );
+
+        $options = array(
+            'upsert' => true,
+        );
+
+        $key->reduceLifetime($this->initialTtl);
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
+        } catch (\MongoDB\Driver\Exception\BulkWriteException $e) {
+            throw new LockConflictedException('Failed to acquire lock', 0, $e);
+        }
+
+        if ($key->isExpired()) {
+            throw new LockExpiredException(sprintf('Failed to store the "%s" lock.', $key));
+        }
+    }
+
+    public function waitAndSave(Key $key)
+    {
+        throw new InvalidArgumentException(sprintf(
+            'The store "%s" does not supports blocking locks.',
+            __CLASS__
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * db.lock.update(
+     *     {
+     *         _id: "test",
+     *         token: {# unique token #},
+     *         expires_at: {
+     *             $gte : new Date()
+     *         }
+     *     },
+     *     {
+     *         _id: "test",
+     *         expires_at: new Date({# now + ttl #})
+     *     },
+     *     {
+     *         upsert: 1
+     *     }
+     * );
+     */
+    public function putOffExpiration(Key $key, $ttl)
+    {
+        $expiry = $this->createDateTime(microtime(true) + $ttl);
+
+        $filter = array(
+            $this->options['resource_field'] => (string) $key,
+            $this->options['token_field'] => $this->getToken($key),
+            $this->options['expiry_field'] => array(
+                '$gte' => $this->createDateTime(),
```

Comment:
I'd rather define a `now` variable at the beggining of the method, then use this variable as a reference for every computes